File: Stacks/Using-LinkedLists/C++/Stack.cpp
#include<bits/stdc++.h>
using namespace std;

//Lo primero que necesitamos hacer es crear la estructura nodo
//Para esto crearemos una clase llamada Node, la cual contendra los atributos que contiene un nodo: El valor y el puntero al siguiente nodo
class Node{
    public:
        int data;
        Node *next;
        Node(){this->data = 0 ; this->next = NULL;}
        Node(int data){this->data = data ; this->next = NULL;}
};

//Ahora crearemos la estructura Lista , en este caso, Stack
//Si bien en la seccion de listas veiamos que esta estructura tenia un nodo cabeza, en este caso para las pilas, lo llamaremos el nodo top el cual se ira acomodando en el ultimo elemento en insertar de la lista
class Stack{
    public:
        Node *top;
        int size;
        Stack(){this->top = new Node() ; this->size = 0;}
};

//Ahora veremos la funcion de insertar en la pila, es decir, la funcion push
//Complejidad Temporal: O(1)
void push(Stack *S ,Node *&top, int data){
    //Creamos el nuevo nodo a insertar
    Node *new_node = new Node(data);
    //Ahora lo que debemos hacer, es que el nuevo nodo creado, apunte al nodo que apunta top
    new_node->next = top->next;
    //Ahora, ya que hay un nuevo nodo en el Stack, debemos de actualizar el puntero de Top ya que este siempre debe de apuntar al nodo que esta en la cima de la pila, es decir, el ultimo nodo en ser insertado
    top->next = new_node;
    //Incrementamos el valor del tamano del Stack
    S->size++;
}

//Ahora veremos como retirar el elemento de la pila, es decir, la funcion pop
//Como se vio en la teoria, en una pila, solo se puede ir sacando desde el dato mas es la cima\
//Complejidad Temporal: O(1)
void pop(Stack *S , Node *&top){
    //Verificamos que el Stack no este vacio
    if(S->size > 0){
        //Como bien sabemos, el nodo top , siempre estara apuntando al nodo en la cima de la pila, es decir, al nodo top
        //Por ello, debemos de actualizar el valor de top, ya que debe de apuntar al siguiente del que esta apuntando
        //Para empezar, guardaremos en un nodo auxiliar, el nodo en la cima de la pila
        Node *peek = top->next;
        //Ya guardado el nodo de la cima, haremos que ahora el nodo top apunte al siguiente del nodo de la cima
        top->next = peek->next;
        //Desvinculamos el que era el nodo de la cima del stack haciendolo apuntar a nulo
        peek->next = NULL;
        //Ahora bien, con el nodo que estaba en la cima del Stack, hay dos opciones, retornar este para utilizarlo en alguna operacion o eliminarlo directamente de memoria. 
        //Esto dependera de que se quiera hacer con este, ya sea solo expusarlo de la pila y  manipularlo, o directamente eliminarlo
        //En este caso eliminaremos el nodo
        delete peek;
        //Decrementamos el vamor de la pila
        S->size--;
    }   
}
 
//Ahora veremos una funcion la cual retornara cual es el dato que esta en la cima de la pila
//En este caso, como estamos trabajando con nodos, podemos decidir si retornar el nodo en donde se encuentra el dato o directamente el dato que se encuentra en el nodo
//Para este caso, se retornara el dato del nodo
//Complejidad Temporal : O(1)
int peek(Stack *S, Node*&top){
    //Verificamos que la pila no este vacia
    if(S->size > 0){
        //Retornamos el dato del nodo que esta en la cima del Stack
        return top->next->data;
    }
}

//Finalmente, haremos la funcion que permite saber si la pila esta vacia o no 
//Complejidad Temporal: O(1)
bool isEmpty(Stack *S){
    //Retorna el valor booleano de la siguiente condicion, indicando si es verdadero o falso
    return S->size > 0;
}


//--------------------------------------------------------------

File: Stacks/Using-LinkedLists/Python/Stack.py
#Lo primero que debemos de hacer para crear un stack utilizando listas , es crear la estructura nodo
#Como se vio en la seccion de listas enlazadas, cada nodo contiene el dato a almcanear y el puntero al siguiente nodo
class Node:
    def __init__(self):  #Constructor vacio
        self.data  = 0 
        self.next = None
    
    def __init__(self,data):  #Constructor que recibe el dato que contendra el nodo
        self.data  = data
        self.next = None

#Ahora debemos de crear la estructura que llamabamos lista en aquella seccion, la cual contenia el nodo cabeza de la lista
#En este caso, llamaremos a esta estructura Stack, y en lugar de hacer un nodo encabezador , lo llamaremos un nodo top el cual siempre apuntara al ultimo nodo en entrar a la pila
class Stack:
    def __init__(self):
        self.top = Node(None)
        self.size = 0

    #Funcion para insertar en un Stack, es decir, la funcion push
    #Complejidad Temporal: O(1)
    def push(self, data):
        #Creamos el nuevo nodo a ser insertado
        new_node = Node(data)
        #Ahora, debemos de hacer que este nuevo nodo apunte al nodo Top del Stack, para ello, crearemos un nodo auxiliar que guarde la referencia inicial del nodo Top
        aux = self.top
        #Con la copia anterior creada, haremos que el nuevo nodo apunte a esta
        new_node.next = aux.next
        #Ahora, como el nodo top siempre debe de apuntar al ultimo nodo en entrar , haremos eso 
        self.top.next = new_node
        #Incrementamos el tamano del Stack
        self.size += 1
    
    #Funcion para retirar elemento de la pila. Por definicion de la pila, solo se puede retirar el elemento que esta en la cima de la pila, es decir, el ultimo entrar 
    #Complejidad Temporal: O(1)
    def pop(self):
        #Verificamos que el Stack no este vacio
        if self.size > 0 :
            #Recordemos que en las pilas, solo se puede retirar el elemento de la cima de esta, no se puede retirar en la posicion que uno desee
            #Para esto, sabemos que el nodo top siempre apunta a este nodo que esta en la cima que necesitamos retirar
            #Como este sera retirado, el nodo top debera de apuntar al nuevo nodo que ocupe la cima, es decir, apuntara al siguiente del nodo que esta en la cima actualmente
            #Primero lo que haremos sera guardar una referencia del nodo que se encuentra actualmente en la cima
            peek = self.top.next
            #Ahora, haremos que el nodo Top apunte a la nueva cima, es decir, el siguiente del nodo de la cima
            self.top.next = peek.next
            #Para desvincular totalmente el nodo que estaba en la cima, lo haremos apuntar a nulo
            peek.next = None
            #Ahora, se pueden realizar dos opciones, se puede retornar este nodo con su dato, o directamente su dato, o se elimina directamente de memoria.
            #Esto dependera de lo que se desee hacer con el nodo expulsado de la pila
            #En este caso se eliminara de memoria
            del peek
            #Decrementamos el valor de la longitud de la pila
            self.size -= 1
    
    #Funcion para saber cual es el elemento que esta en la cima de la pila
    #Complejidad Temporal: O(1)
    def peek(self):
        #Verificamos que la lista no este vacia
        if self.size > 0 :
            #Aca como se esta trabajando con nodos, se puede retornar el nodo o directamente el dato que contiene
            #En este caso, retornaremos el dato que contiene el nodo
            return self.top.next.data
    
    #Funcion para determinar si la pila esta vacia o no
    #Complejidad Temporal: O(1)
    def isEmpty(self):
        #Retornamos el valor booleano del condicional 
        return self.size > 0


#------------------------------------------------------------------------------------

